oprf/group/suite.go

```diff
@@ -0,0 +1,251 @@
+package group
+
+import (
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"io"
+	"math/big"
+
+	"github.com/armfazh/h2c-go-ref"
+)
+
+// Ciphersuite corresponds to the OPRF ciphersuite that is chosen. The
+// Ciphersuite object determines the prime-order group (pog) that is used for
+// performing the OPRF operations, along with the different hash function
+// definitions.
+// Should be created using NewSuite, using an string.
+type Ciphersuite struct {
+	// name of the ciphersuite.
+	name string
+
+	// dst is a tag to be used for hashing to curve.
+	dst []byte
+
+	// hash defines the hash function to be used.
+	hash hash.Hash
+
+	curve elliptic.Curve
+}
+
+// Suite is an interface that defines operations within additive
+// groups of prime order. This is the setting in which the OPRF operations
+// take place.
+type Suite interface {
+	// Returns the identifying name of the group
+	Name() string
+
+	// Returns the canonical (fixed) generator for defined group
+	Generator() *Element
+
+	// Returns the order of the canonical generator in the group.
+	Order() *Scalar
+
+	// Performs a transformation to encode bytes as a Element object in the
+	// group.
+	HashToGroup([]byte) (*Element, error)
+
+	// Performs a transformation to encode bytes as a Scalar object in the
+	// appropriate group.
+	HashToScalar([]byte) (Scalar, error)
+
+	// Samples a random scalar value from the field of scalars defined by the
+	// group order.
+	RandomScalar() *Scalar
+
+	// Performs a scalar multiplication of the Generator with some scalar
+	// input
+	ScalarMultBase(*Scalar) *Element
+
+	// Returns the ByteLength of objects associated with the Ciphersuite
+	ByteLength() int
+}
+
+// Name returns the name associated with the chosen ciphersuite.
+func (c *Ciphersuite) Name() string {
+	return c.name
+}
+
+// Generator returns the canonical (fixed) generator for the defined group.
+func (c *Ciphersuite) Generator() *Element {
+	return &Element{c, c.curve.Params().Gx, c.curve.Params().Gy, true}
+}
+
+// Order returns the order of the canonical generator in the group.
+func (c *Ciphersuite) Order() *Scalar {
+	return &Scalar{c, c.curve.Params().N}
+}
+
+func getH2CSuite(c *Ciphersuite) (HashToPoint, error) {
+	var suite h2c.SuiteID
+
+	switch c.Name() {
+	case "OPRFP256-SHA512-ELL2-RO":
+		suite = h2c.P256_XMDSHA256_SSWU_RO_
+	case "OPRFP384-SHA512-ELL2-RO":
+		suite = h2c.P384_XMDSHA512_SSWU_RO_
+	case "OPRFP521-SHA512-ELL2-RO":
+		suite = h2c.P521_XMDSHA512_SSWU_RO_
+	default:
+		return nil, errors.New("invalid suite")
+	}
+
+	hasher, err := suite.Get(c.dst)
+	if err != nil {
+		return nil, err
+	}
+
+	return eccHasher{c, hasher}, nil
+}
+
+// HashToPoint produces a new point by hashing the input message.
+type HashToPoint interface {
```

Comment:
```suggestion
type HashToElement interface {
```

(To align with the Element terminology)

Reply:
No, it needs to be called like that as the h2c requires that name for the interface.

Reply:
Sorry, my bad. I thought this was something else. Changed now.